plugins/node/opentelemetry-plugin-koa/test/koa.test.ts

```diff
@@ -0,0 +1,217 @@
+  provider.addSpanProcessor(spanProcessor);
+  const tracer = provider.getTracer('default');
+  let contextManager: AsyncHooksContextManager;
+
+  before(() => {
+    plugin.enable(koa, provider, logger);
+  });
+
+  beforeEach(() => {
+    contextManager = new AsyncHooksContextManager();
+    context.setGlobalContextManager(contextManager.enable());
+  });
+
+  afterEach(() => {
+    memoryExporter.reset();
+    context.disable();
+  });
+
+  const simpleResponse: koa.Middleware = (ctx, next) => {
+    ctx.body = 'test';
+    return next();
+  };
+
+  const customMiddleware: koa.Middleware = (ctx, next) => {
+    for (let i = 0; i < 1000000; i++) {
+      continue;
+    }
+    return next();
+  };
+
+  const asyncMiddleware: koa.Middleware = async (ctx, next) => {
+    const start = Date.now();
+    await next();
+    const ms = Date.now() - start;
+    console.log(`${ctx.method} ${ctx.url} - ${ms}ms`);
+  };
+
+  describe('Instrumenting core middleware calls', () => {
+    it('should create a child span for middlewares', async () => {
+      const rootSpan = tracer.startSpan('rootSpan');
+      const app = new koa();
+      app.use((ctx, next) => tracer.withSpan(rootSpan, next));
+      app.use(customMiddleware);
+      app.use(simpleResponse);
+
+      app.use((ctx, next) => {
+        ctx.body = 'anonymous';
+        return next();
+      });
+
+      const server = http.createServer(app.callback());
+      await new Promise(resolve => server.listen(0, resolve));
+      const port = (server.address() as AddressInfo).port;
+      assert.strictEqual(memoryExporter.getFinishedSpans().length, 0);
+
+      await tracer.withSpan(rootSpan, async () => {
+        await httpRequest.get(`http://localhost:${port}`);
+        rootSpan.end();
+
+        assert.notStrictEqual(
+          memoryExporter
+            .getFinishedSpans()
+            .find(span => span.name.includes('customMiddleware')),
+          undefined
+        );
+
+        const simpleResponseSpan = memoryExporter
+          .getFinishedSpans()
+          .find(span => span.name.includes('simpleResponse'));
+        assert.notStrictEqual(simpleResponseSpan, undefined);
+
+        const requestHandlerSpan = memoryExporter
+          .getFinishedSpans()
+          .find(span => span.name.includes('middleware'));
+        assert.notStrictEqual(requestHandlerSpan, undefined);
+        assert.strictEqual(
+          requestHandlerSpan?.attributes[AttributeNames.COMPONENT],
+          KoaComponentName
+        );
+        assert.strictEqual(
+          requestHandlerSpan?.attributes[AttributeNames.KOA_TYPE],
+          KoaLayerType.MIDDLEWARE
+        );
+        const exportedRootSpan = memoryExporter
+          .getFinishedSpans()
+          .find(span => span.name === 'rootSpan');
+        assert.notStrictEqual(exportedRootSpan, undefined);
+      });
+      server.close();
+    });
+
+    it('should not create span if there is no parent span', async () => {
```

Comment:
Out of curiosity, was this an intentional design decision? 
- I am assuming that `koa.test` line=120 is the place that checks for it (please correct me if my understanding is not correct).
If the user does not explicitly create a starting/root span, will there me no middleware/routes spans?

Reply:
Yes, that was intentional and that is the line which checks this! This design is used in the Express and MongoDB plugins as well, which is why I implemented it. For some reason, it does not seem to be checked in some other plugins though, which is a bit odd. I will ask some of the OTel contributors for their input once I make the public PR

Reply:
Thank you for the explanation!